lib/bundler/definition.rb

```diff
@@ -480,14 +482,21 @@ def specs_changed?(source, &block)
         end
       end
 
-      !locked || unlocking || dependencies_for_source_changed?(locked) || source.specs != locked.specs
+      !locked || unlocking || dependencies_for_source_changed?(source) || specs_for_source_changed?(source)
```

Comment:
`source.specs != locked.specs` is the key problem.

`source.specs` takes our source of gemspecs and enumerates the specs for that source by finding all the gemspecs for that source from the filesystem, etc.

`locked.specs` takes our Gemfile.lock version of this same source and calls specs, which does the SAME THING.  It finds the gemspecs for the location specified from the Gemfile.lock, which is the same as the source, and gets the exact same list.  This doesn't actually produce the list of gemspecs from our lockfile.